st2actions/st2actions/workflows/workflows.py

```diff
@@ -85,6 +93,53 @@ def process(self, message):
             # the database and fail the workflow execution gracefully. In this case,
             # the garbage collector will find and cancel these workflow executions.
             self.fail_workflow_execution(message, e)
+        finally:
+            if type(message) == wf_db_models.WorkflowExecutionDB:
+                self._running_workflows.remove(str(message["action_execution"]))
+            else:
+                try:
+                    self._running_workflows.remove(str(message["parent"]))
+                except ValueError:
+                    pass
+
+    def shutdown(self):
+        super(WorkflowExecutionHandler, self).shutdown()
+
+        if cfg.CONF.workflow_engine.graceful_shutdown:
+
+            coordinator = coordination.get_coordinator()
+            member_ids = []
+            service = "workflow_engine"
+            exit_timeout = cfg.CONF.workflow_engine.exit_timeout
+            sleep_delay = cfg.CONF.workflow_engine.sleep_delay
+            timeout = 0
+
+            while timeout < exit_timeout and self._running_workflows:
+                try:
+                    member_ids = list(
+                        coordinator.get_members(service.encode("utf-8")).get()
+                    )
+                except GroupNotCreated:
+                    pass
+
+                # Check if there are other runners in service registry
+                if not member_ids:
+                    break
+                timeout += sleep_delay
+                concurrency.sleep(sleep_delay)
+
+            running_workflows = set(self._running_workflows)
+
+            while running_workflows:
+                execution_id = running_workflows.pop()
+                try:
+                    ac_ex_db = ex_db_access.ActionExecution.get_by_id(str(execution_id))
+                    lv_ac_db = lv_db_access.LiveAction.get_by_id(
+                        ac_ex_db.liveaction["id"]
+                    )
```

Comment:
Can you create a common function get_liveaction_by_execution in https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/action.py? I thought there may be some common function like this already but I couldn't find any. Please add appropriate unit test(s) as well.